js/pathfinding.js

```diff
@@ -5,56 +5,159 @@ import {settingsKey} from "./settings.js";
-import {PriorityQueueSet} from "./data_structures.js";
+import {PriorityQueueSet, ProcessOnceQueue} from "./data_structures.js";
 
 class Cache {
 	static maxCacheIds = 5;
+	static maxBackgroundCachingMillis = 10;
 
 	constructor() {
 		this.nodes = new Map();
 		this.lastUsed = new Map();
+		this.background = {
+			queues: new Map(),
+			nextJobId: null
+		}
 	}
 
 	clear() {
 		this.nodes.clear();
 		this.lastUsed.clear();
+		this.background.queues.clear();
+		if (this.background.nextJobId) {
+			window.cancelIdleCallback(this.background.nextJobId);
+			this.background.nextJobId = null;
+		}
 	}
 
 	/**
-	 * Get the cache associated with the given cache ID, creating a new one
-	 * if we don't already have one
+	 * Build a cache ID from the information that could make a difference to the pathfinding algorithm
 	 */
-	getCachedNodes(cacheId) {
+	getCacheId(token) {
+		const cacheData = {};
+
+		// Different-sized tokens snap to different points on the grid,
+		// so they might follow a different path to other tokens
+		cacheData.tokenSize = getTokenSize(token);
+		if (canvas.grid.isHex && game.modules.get("hex-size-support")?.active) {
+			cacheData.hexConfig = {
+				altOrientation: CONFIG.hexSizeSupport.getAltOrientationFlag(token),
+				altSnapping: CONFIG.hexSizeSupport.getAltSnappingFlag(token)
+			}
+		}
+
+		// If levels is enabled, the token's elevation can affect which walls
+		// they need to worry about
+		if (game.modules.get("levels")?.active) {
+			cacheData.elevation = token.data.elevation;
+		}
+
+		const cacheId = JSON.stringify(cacheData);
+		this.ensureCacheExists(cacheId);
+		return cacheId;
+	}
+
+	ensureCacheExists(cacheId) {
 		// Track that we've last used this cache right now
 		this.lastUsed.set(cacheId, Date.now());
 
 		// Get the nodes for the cacheId. If we don't already have one, create one
-		let cachedNodes = this.nodes.get(cacheId);
-		if (!cachedNodes) {
-			cachedNodes = new Array(gridHeight);
+		if (!this.nodes.has(cacheId)) {
+			const cachedNodes = new Array(gridHeight);
 			for (let y = 0; y < gridHeight; y++) {
 				cachedNodes[y] = new Array(gridWidth);
 				for (let x = 0; x < gridWidth; x++) {
 					cachedNodes[y][x] = {x, y};
 				}
 			}
 			this.nodes.set(cacheId, cachedNodes);
+			this.background.queues.set(cacheId, new ProcessOnceQueue());
 
 			// Since we're adding a new cache, check if we have too many and,
 			// if we do, get rid of the one that was last used longest ago
 			if (this.lastUsed.size > Cache.maxCacheIds) {
-				let oldest;
-				for (let entry of this.lastUsed) {
-					if (!oldest || oldest[1] > entry[1]) {
-						oldest = entry;
+				// Build an array from the last used entries, sort in ascending order, and get the first
+				// element (with the lowest last-used value) as the oldest entry, then retrieve the cache ID
+				// from that element
+				const oldestCacheId = Array.from(this.lastUsed.entries()).sort((a, b) => a[1] - b[1])[0][0];
```

Comment:
Since this is going to be a maximum of 5 elements, it's cleaner to make it into an array, sort it, and get the first one.

Reply:
You could use `reduce` instead of `sort` to get the smallest element in `O(n)` runtime complexity with roughly the same code complexity (not that it mattered too much with 5 elements, but we might want to raise that limit in the future).